[VectorCombine] Fix invalid shuffle cost argument of foldShuffleOfSelects

ParkHanbum opened the pull request:
In the previous code, it specified the destination vector as the getShuffleCost argument. Because the shuffle mask specifies the indices of the two vectors specified as elements, the maximum value is twice the size of the source vector. This causes a problem if the destination vector is smaller than the source vector and specify an index in the mask that exceeds the size of the destination vector.

Fix the problem by correcting the previous code, which was using wrong argument in the Cost calculation.

Fixed https://github.com/llvm/llvm-project/issues/130250

ParkHanbum committed:
[VectorCombine] Fix invalid shuffle cost argument of foldShuffleOfSelects

In the previous code, it specified the destination vector as
the getShuffleCost argument. Because the shuffle mask specifies
the indices of the two vectors specified as elements, the maximum
value is twice the size of the source vector. This causes a problem
if the destination vector is smaller than the source vector and
specify an index in the mask that exceeds the size of the destination
vector.

Fix the problem by correcting the previous code, which was using
wrong argument in the Cost calculation.

Fixed https://github.com/llvm/llvm-project/issues/130250

RKSimon commented on llvm/lib/Transforms/Vectorize/VectorCombine.cpp:
```diff
@@ -2051,24 +2050,26 @@ bool VectorCombine::foldShuffleOfSelects(Instruction &I) {
-      SelOp, T1->getType(), C1VecTy, CmpInst::BAD_ICMP_PREDICATE, CostKind);
-  OldCost += TTI.getCmpSelInstrCost(SelOp, T2->getType(), C2VecTy,
+      SelOp, DstVecTy, C1VecTy, CmpInst::BAD_ICMP_PREDICATE, CostKind);
+  OldCost += TTI.getCmpSelInstrCost(SelOp, DstVecTy, C2VecTy,
```
Should still be SrcVecTy for the getCmpSelInstrCost calls?

ParkHanbum replied:
I see that SelectInst takes a DestVecTy as an argument, see VPlanRecipes.cpp:2192.
am I wrong?

ParkHanbum replied:
ahhh..! you right!

ParkHanbum replied:
I've fixed the my stupid mistake, thank you.

ParkHanbum committed:
Fix invalid argument in getCmpSelInstrCost

The target vector for SelectInst is SrcVecTy. It was a mistake
to specify DstVecTy as an argument before. it is now fixed.

RKSimon commented on llvm/test/Transforms/PhaseOrdering/X86/blendv-select.ll:
```diff
@@ -12,10 +12,20 @@
+; AVX512-LABEL: @x86_pblendvb_v4f64_v2f64(
+; AVX512-NEXT:    [[CMP:%.*]] = fcmp olt <4 x double> [[C:%.*]], [[D:%.*]]
+; AVX512-NEXT:    [[DOTV:%.*]] = select <4 x i1> [[CMP]], <4 x double> [[B:%.*]], <4 x double> [[A:%.*]]
+; AVX512-NEXT:    ret <4 x double> [[DOTV]]
```
Regenerate this file?

RKSimon reviewed (approved):
LGTM
